service/models.py

```diff
@@ -267,12 +233,107 @@ def find(cls, by_id):
+    def filter_by_datetime(cls, datetime, query):
+        """Returns all promotions which are valid at the specified datetime
+
+        Args:
+            datetime (_type_): datetime where promotion should be valid
+        """
+        return query.filter(and_(cls.start_date <= datetime, cls.end_date >= datetime))
+
+    @classmethod
+    def filter_by_promotion_type(cls, promotion_types, query):
+        """Returns all promotions which have the specified type
+
+        Args:
+            promotion_type (PromotionType): promotion type to match
+        """
+        return query.filter(cls.promotion_type.in_(promotion_types))
+
+    @classmethod
+    def filter_by_promotion_scope(cls, promotion_scopes, query):
+        """Returns all promotions which have the specified scope
+
+        Args:
+            promotion_scope (PromotionScope): promotion scope to match
+        """
+        return query.filter(cls.promotion_scope.in_(promotion_scopes))
+
     @classmethod
     def find_by_name(cls, name):
-        """Returns all YourResourceModels with the given name
+        """Returns all Promotions with the given name
 
         Args:
-            name (string): the name of the YourResourceModels you want to match
+            name (string): the name of the Promotions you want to match
         """
         logger.info("Processing name query for %s ...", name)
         return cls.query.filter(cls.promotion_name == name)
+
+    @classmethod
+    def deserialize_with_default(cls, key, data, default, deserializer=None):
+        """Deserializes a field with a provided key from incoming json with a default value.
+        Uses provided deserializer if provided
+
+        Args:
+            key (string): The string to extract from the data
+            data (dict): The JSON object containing the incoming data
+            default (Any): the default to use if no data is found
+            deserializer (function, optional): Provided deserializer. Defaults to None.
+        """
+        if not data.get(key, None):
+            return default
+        if deserializer is not None:
+            return deserializer(data.get(key))
+        return data.get(key)
+
+    @classmethod
+    def deserialize_datetime(cls, datetime_str: str):
+        """
+        Deserialize a datetime from a datetime string
+        Args:
+            datetime_str: A string representing the date
+        """
+        try:
+            return datetime_from_str(datetime_str)
+        except ValueError as error:
+            raise DataValidationError(
+                f"Invalid date format: {datetime_str} does not conform to any valid datetime format"
+            ) from error
+
+    @classmethod
+    def to_list_deserializer(cls, deserializer):
+        """Converts a deserialization function to an equivalent one for a list of the specified object"""
+
+        def deserialize_list(ls):
+            return [deserializer(item) for item in ls]
+
+        return deserialize_list
```

Comment:
How about putting these three methods at the top of the CLASS METHODS section? Several new functions will be implemented afterward.

Reply:
There is no difference, just because there are several functions named with find_by_xxx in the subsequent stories, I think putting these functions together will keep our code more organized. Thank you for the help! : )

Reply:
Makes complete sense! Thanks :)